RNDF Utility functions to work with Bezier, Splines and PChip curves.

agalbachicar opened the pull request:
Since piecewise splines do not provide guarantee monotonicity nor convexity at each piece, this pull request provides some utility functions that adjust splines geometries to that end for the purpose of creating smooth roads.

Cubic splines can be converted to cubic Bezier curves whose control points can be easily adjusted to prevent undesired shapes while preserving headings at the ends. This becomes useful when deriving the geometry that connects two Maliput's `BranchPoints` (`exits` and `entries` in RNDF spec). 

In addition, since [PChip](https://github.com/RobotLocomotion/drake/blob/master/drake/common/trajectories/piecewise_polynomial.h#L100-L139) guarantees monotonicity and convexity at each piece, creating an spline from PChip's knots provides the desired geometry with Maliput's required characteristics. This is convenient when creating a curve from a set of RNDF waypoints.

This code will be consumed by the RNDF builder in a future PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6569)
<!-- Reviewable:end -->

agalbachicar commented:
+@liangfok for feature review please.

agalbachicar commented:
> +@liangfok for feature review please.
+@liangfok for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KoxDfjEABG25Rekup8O:b7mlb05)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@liangfok for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KoxE6vO2ud4Jgw57L8Q:b-pbut8i)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed first review pass.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLJK0d2fiTREfSi5t3:-KpLJK0d2fiTREfSi5t4:brq82ue) (no related file):*
I just realized that the Doxygen documentation for spline_helpers.h is not showing up when built using Bazel. Can you figure out why?

http://drake.mit.edu/documentation_instructions.html#documentation-generation-instructions

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpKyNS-LKWL-gjVAanJ:-KpKyNS-LKWL-gjVAanK:b-5tjy42) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L38)):*
The above two matrices violate the following style guide rule:

> Variables of class type with static storage duration are forbidden.

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Consider something like the following:

```c++
static constexpr double kBezierMatrix[16] = {
   1.0,  0.0,  0.0, 0.0,
  -3.0,  3.0,  0.0, 0.0,
   3.0, -6.0,  3.0, 0.0,
  -1.0,  3.0, -3.0, 1.0};
```

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 267 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL9j0E-4ZpWtIFhM8o:-KpL9j0E-4ZpWtIFhM8p:b-5tt6x6) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L267)):*
> ```cpp
>   // [hermite_points] =
>   //     [kHermiteMatrix]^(-1) * [kBezierMatrix] * [bezier_points]
>   ignition::math::Matrix4d hermite_points =
> ```

BTW, `const`?

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 287 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLA1JQvNdlSvtVslI1:-KpLA1JQvNdlSvtVslI2:bn3672j) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L287)):*
> ```cpp
>   DRAKE_THROW_UNLESS(scale <= 1.0);
>   DRAKE_THROW_UNLESS(scale >= 0.0);
>   // Sets a constant to compare cross product results near to zero.
> ```

BTW, consider deleting "to" before "zero".

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLAQmAIhaJFw_cbwYH:-KpLAQmBZby4Loxzi54V:b-i6c64x) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L307)):*
> ```cpp
>   if (norm_t3_x_r.Length() < kAlmostZero ||
>       norm_t3_x_norm_t0.Length() < kAlmostZero) {
>     // Lines are not coplanar or are parallel so it will not intersect. As RNDF
> ```

BTW, "it" should be "they".

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLAYYvtYMxa5aViu58:-KpLAYYvtYMxa5aViu59:b12x596) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L309)):*
> ```cpp
>     // Lines are not coplanar or are parallel so it will not intersect. As RNDF
>     // does not support any other z coordinate different from 0.0, it is assumed
>     // that lines are coplanar and parallel. Consequently, control points for
> ```

"the resulting Bezier curve" should be added after "for".

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 144 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL2rROnqAytrqLkHIj:-KpL2rROnqAytrqLkHIk:b-bnug2t) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L144)):*
> ```objc
> };
> 
> /// Provides the equivalent set of points in cubic Bezier base from a two pairs
> ```

Delete "a".

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL2uFH_LAdSoOU4rOU:-KpL2uFH_LAdSoOU4rOV:b-l7kgcw) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L145)):*
> ```objc
> 
> /// Provides the equivalent set of points in cubic Bezier base from a two pairs
> /// of point and tangents at the extents of a spline.
> ```

"point" should be "points".

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL3GW3a9WCxaizZThy:-KpL3GW3a9WCxaizZThz:bs5mfs2) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L164)):*
> ```objc
> /// @param p3 A vector that describes the last control point of the curve.
> /// @return A vector containing four spline control points. The points are
> /// return in the following order:
> ```

"return" should be "returned".

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL3oYfUe27f6B5hN02:-KpL3oYgBmMcwRotXs-h:blq1t18) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L179)):*
> ```objc
> /// a first branch in the behavior, if there is no intersection, we assume that
> /// all these curves are 2-D curves over the z = 0.0 api::GeoPosition frame,
> /// that lines are parallel. Then, we create two intermediate control points for
> ```

BTW, consider deleting "that lines are parallel" since that's already implied by the fact that there is no intersection.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 182 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL4Qvbrx9GSP_jO6qw:-KpL4Qvbrx9GSP_jO6qx:b-321okh) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L182)):*
> ```objc
> /// them that will provide a S shape to match the curve. The change in convexity
> /// is set to be in the mid point of the extents of the curve.
> /// In case there is an intersection in the curve, we find it and this will be
> ```

Should "in the curve" be "between the two lines"?

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 187 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL4KIRh3KPJtxspDqz:-KpL4KIRh3KPJtxspDr-:b-jjgk7w) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L187)):*
The following sentence is somewhat difficult to follow:

> When this happens we assign as control points the beginning of the curve, the beginning plus a difference vector scaled by @p scale, the ending point minus another difference vector scaled by @p scale and finally the ending point.

Maybe conveying it through math equations will make it easier to understand?

If `@p scale` is > 0.5, doesn't that place the two control points closer to the opposite end points? Is this OK?

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 199 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL5S0IOJKxnbO0YO23:-KpL5S0IOJKxnbO0YO24:bc0x5px) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L199)):*
> ```objc
> /// @throws when the size of @p control_points is different from 4.
> /// @throws when @p scale is bigger than 1.0.
> /// @throws when @p scale is smaller than 0.0.
> ```

BTW, I believe the @throws notation requires specifying the exception type:  http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdthrows.

(Same comment below.)

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCgdACUCoTT8aPtab:-KpLCgdACUCoTT8aPtac:b-b1uv6a) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L171)):*
BTW, consider replacing:

> shape a 90° figure

with:

> form a 90° shape

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 193 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCwzcySYsvMoP24A5:-KpLCwzcySYsvMoP24A6:b-tr9tbx) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L193)):*
> ```cpp
>                                             kLinearTolerance));
> 
>   // Second case sets points under the critical curvature
> ```

BTW, missing period at end of sentence.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 217 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCz3WGn0dHKETK_Dl:-KpLCz3WGn0dHKETK_Dm:b-tr9tbx) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L217)):*
> ```cpp
>                                             kLinearTolerance));
> 
>   // Third case sets points above the critical curvature
> ```

BTW, missing period at end of sentence.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDGxh46IrTs6AJ3U6:-KpLDGxh46IrTs6AJ3U7:bkeqwiu) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L246)):*
> ```cpp
> GTEST_TEST(RNDFBezierToBezierTest, ConversionTest) {
>   ignition::math::Vector3d p0, p1, p2, p3;
>   // First case where the points control points shape a 90° figure. Control
> ```

BTW, delete "points" before "control".

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDNFMrlCAdySB-KzK:-KpLDNFMrlCAdySB-KzL:b-b1uv6a) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L246)):*
BTW, consider replacing:

> shape a 90° figure

with:

> form a 90° shape

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 295 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDYugWgAZpPU4piHN:-KpLDYugWgAZpPU4piHO:b3h0tbj) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L295)):*
> ```cpp
>                                                 0.0),
>                                             kLinearTolerance));
>   // Third case sets points above the critical curvature.
> ```

BTW, consider adding a blank line above this line for consistency.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLFWT9exMMqiZiQBFp:-KpLFWT9exMMqiZiQBFq:b-vx6rvj) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L527)):*
> ```cpp
>   // Checks that it will abort if two consecutive points are the same.
>   points.push_back(ignition::math::Vector3d(20.0, 0.0, 0.0));
>   EXPECT_DEATH(CreatePChipBasedSpline(points), "(.*?)");
> ```

This does not satisfy the death test style:

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#death-test-styles

Consider making it thrown exception that's checked here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpLH7XJigDUwBQvdWAk:bnsr0cf)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I have addressed all the comments. PTAL.

---

Review status: 0 of 4 files reviewed at latest revision, 20 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLJK0d2fiTREfSi5t3:-KpQAf2lI0ardAF8gjbJ:bdiavyv) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I just realized that the Doxygen documentation for spline_helpers.h is not showing up when built using Bazel. Can you figure out why?

http://drake.mit.edu/documentation_instructions.html#documentation-generation-instructions
</blockquote></details>

Sorry, but I cannot find the reason. I have inspected and compared spline_helpers.h against monolane/builder.h which holds some class definitions in the same file and I have not been able to find the difference. In addition, I have separated InverseArcLengthInterpolator class and ArcLengthParametizedSpline class into two different namespace blocks and InverseArcLengthInterpolator class documentation is parsed but the other one is not. The same happens when I moved InverseArcLengthInterpolator class definition to another header file.
Moreover, if I change class comments of ArcLengthParametizedSpline to be something like "/// Class description." it is not parsed. Tthe last comment, if you take a look at the parsed spline_helpers.h file done by Doxygen (from master) (http://drake.mit.edu/doxygen_cxx/spline__helpers_8h_source.html) you will see the file without Doxygen comments.

Do you have any suggestions?

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpKyNS-LKWL-gjVAanJ:-KpLwG_AUofzB2z3tJ53:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L38)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The above two matrices violate the following style guide rule:

> Variables of class type with static storage duration are forbidden.

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Consider something like the following:

```c++
static constexpr double kBezierMatrix[16] = {
   1.0,  0.0,  0.0, 0.0,
  -3.0,  3.0,  0.0, 0.0,
   3.0, -6.0,  3.0, 0.0,
  -1.0,  3.0, -3.0, 1.0};
```
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 267 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL9j0E-4ZpWtIFhM8o:-KpLwJlq_t0G7T7-2z69:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L267)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, `const`?
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 287 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLA1JQvNdlSvtVslI1:-KpLwOq7XD0swXg_8WGS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L287)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider deleting "to" before "zero".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLAQmAIhaJFw_cbwYH:-KpLwYQ6fsM517flPMRB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L307)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "it" should be "they".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.cc, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLAYYvtYMxa5aViu58:-KpLwh6hrXRhYtMRyrbl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.cc#L309)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"the resulting Bezier curve" should be added after "for".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 144 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL2rROnqAytrqLkHIj:-KpLwqzBo5Kdk4oowZF_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L144)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Delete "a".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL2uFH_LAdSoOU4rOU:-KpLwuGf50Eoo2Ru4sb3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L145)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"point" should be "points".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL3GW3a9WCxaizZThy:-KpLx-hqfADJHKscxFp-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L164)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

"return" should be "returned".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL3oYfUe27f6B5hN02:-KpLxQeqqG0FzIfTaIpW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L179)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider deleting "that lines are parallel" since that's already implied by the fact that there is no intersection.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 182 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL4Qvbrx9GSP_jO6qw:-KpLxfr5j_QWML6pWKwD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L182)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should "in the curve" be "between the two lines"?
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 187 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL4KIRh3KPJtxspDqz:-KpM77Z-kdBiKN8FHcXl:bsjq8l5) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L187)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The following sentence is somewhat difficult to follow:

> When this happens we assign as control points the beginning of the curve, the beginning plus a difference vector scaled by @p scale, the ending point minus another difference vector scaled by @p scale and finally the ending point.

Maybe conveying it through math equations will make it easier to understand?

If `@p scale` is > 0.5, doesn't that place the two control points closer to the opposite end points? Is this OK?
</blockquote></details>

Done. It depends. The idea is that control_points in the middle will be located over the segment that joins the extent control points and the critical point. So, scale is just a factor that lets you put the control points in between from a coincident position to the first or last control point (scale is 0) to both control points at the critical control point (scale = 1). I guess the following issue will give a better understanding of how this method work: https://bitbucket.org/ekumen/terminus-simulation/issues/170/curves-research-for-rndf-maliput

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 199 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpL5S0IOJKxnbO0YO23:-KpLy7QiF4BTmgjaBSGG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/spline_helpers.h#L199)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I believe the @throws notation requires specifying the exception type:  http://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdthrows.

(Same comment below.)
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCgdACUCoTT8aPtab:-KpLyKxpFF5xyqci_yh7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L171)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing:

> shape a 90° figure

with:

> form a 90° shape
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 193 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCwzcySYsvMoP24A5:-KpLyV7XtzfqvFRsA3MC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L193)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing period at end of sentence.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 217 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLCz3WGn0dHKETK_Dl:-KpLyV_1JpExRTnZKnNo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L217)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing period at end of sentence.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDGxh46IrTs6AJ3U6:-KpLyhJRetge-8cc9KwX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L246)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, delete "points" before "control".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 246 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDNFMrlCAdySB-KzK:-KpLyiNLsKmjfnn1L06p:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L246)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing:

> shape a 90° figure

with:

> form a 90° shape
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 295 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLDYugWgAZpPU4piHN:-KpLymd-IYurtK-sTev5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L295)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a blank line above this line for consistency.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLFWT9exMMqiZiQBFp:-KpLzFtweiMmjxu86SyE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L527)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This does not satisfy the death test style:

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#death-test-styles

Consider making it thrown exception that's checked here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpQDJ84FNhfd6WWyjwL:b6bbsn6)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed second review pass.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLJK0d2fiTREfSi5t3:-KpueKbh7YZIMZwC4J3j:brypx9l) (no related file):*
The following error was recorded in drake-distro/build/drake/doc/doxygen_cxx.log:

> drake-distro/drake/automotive/maliput/rndf/spline_helpers.h:226: warning: File ended in the middle of a comment block! Perhaps a missing \endcode?

I identified the missing endcode and left a comment below to address it.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 112 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpueXywtMeiysP-tJ-c:-KpueXywtMeiysP-tJ-d:b-rfjrd4) ([raw file](https://github.com/robotlocomotion/drake/blob/004298ca24db228fca1d1219d0988ae1b1998afe/drake/automotive/maliput/rndf/spline_helpers.h#L112)):*
> ```objc
>   /// @param[in] s path length to interpolate at, constrained
>   /// by the curve dimensions [0, path_length].
>   /// @return the @p derivative_order derivative @f$ Q^{(derivative_order)}(s) .
> ```

Doxygen generation for this file fails due to a missing closing `@f$`. This line should be:

```
  /// @return the @p derivative_order derivative
  ///         @f$ Q^{(derivative_order)}(s) @f$.
```

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 207 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-Kp_mtD7HUaMeEVfx_ay:-Kp_mtD7HUaMeEVfx_az:b-dx9p6x) ([raw file](https://github.com/robotlocomotion/drake/blob/004298ca24db228fca1d1219d0988ae1b1998afe/drake/automotive/maliput/rndf/spline_helpers.h#L207)):*
> ```objc
> /// @throws std::runtime_error When @p scale is bigger than 1.0.
> /// @throws std::runtime_error When @p scale is smaller than 0.0.
> std::vector<ignition::math::Vector3d> AdjustBezierGeometry(
> ```

BTW, consider modifying the name of this method to be more descriptive. Something like `MakeBezierCurveConvexAndMonotonic`.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLFWT9exMMqiZiQBFp:-Kp_dL8-36fiIPzmwpiG:b-phik0a) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L527)):*
The name of this test case should end with the words "DeathTest".

> We strongly recommend you to follow the convention of naming your test case (not test) *DeathTest when it contains a death test, as demonstrated in the above example.

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#how-to-write-a-death-test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-Kp_ej1qD0S1AIJwyJRl:bq4tjd3)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I have addressed the comments. PTAL.

---

Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLJK0d2fiTREfSi5t3:-KpvTJDGr48L2I64eSu9:b-nwopp3) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The following error was recorded in drake-distro/build/drake/doc/doxygen_cxx.log:

> drake-distro/drake/automotive/maliput/rndf/spline_helpers.h:226: warning: File ended in the middle of a comment block! Perhaps a missing \endcode?

I identified the missing endcode and left a comment below to address it.
</blockquote></details>

Thanks for pointing it out, I've completely missed it. 

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 112 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpueXywtMeiysP-tJ-c:-Kpv6-Ft5eEjKp01PFnl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/004298ca24db228fca1d1219d0988ae1b1998afe/drake/automotive/maliput/rndf/spline_helpers.h#L112)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Doxygen generation for this file fails due to a missing closing `@f$`. This line should be:

```
  /// @return the @p derivative_order derivative
  ///         @f$ Q^{(derivative_order)}(s) @f$.
```
</blockquote></details>

Done.

---

*[drake/automotive/maliput/rndf/spline_helpers.h, line 207 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-Kp_mtD7HUaMeEVfx_ay:-KpvFWlo-bl17P9EtYp2:b-bp1zhi) ([raw file](https://github.com/robotlocomotion/drake/blob/004298ca24db228fca1d1219d0988ae1b1998afe/drake/automotive/maliput/rndf/spline_helpers.h#L207)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider modifying the name of this method to be more descriptive. Something like `MakeBezierCurveConvexAndMonotonic`.
</blockquote></details>

Done. Changed to MakeBezierCurveMonotonic since the result can conditionally make a convex curve.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 527 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpLFWT9exMMqiZiQBFp:-Kpv9iBakR5cS04en2TN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/612624e868ba1c4c9d9cd75c624a6c4c8fac5e53/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L527)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

The name of this test case should end with the words "DeathTest".

> We strongly recommend you to follow the convention of naming your test case (not test) *DeathTest when it contains a death test, as demonstrated in the above example.

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#how-to-write-a-death-test
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpvTFWPBb2fS8mlfsGi:bg1u43l)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpvlsQ8J-v0x_b9X8p3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
+@ggould-tri for platform review.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-Kpz8fic0U8t-oFdb0Ty:b-8qo47q)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for one item around death testing, below.

---

Reviewed 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 519 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpzWoCzPazTlFJO87ed:-KpzWoCzPazTlFJO87ee:brld7i2) ([raw file](https://github.com/robotlocomotion/drake/blob/380660a03d5fa529bebfeccdbaf63f9189ac96ff/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L519)):*
> ```cpp
> // Checks that all the constraints throw when conditions are not satisfied.
> GTEST_TEST(RNDFCreatePChipBasedSplineTestDeathTest, CaseExceptions) {
>   ::testing::FLAGS_gtest_death_test_style = "fast";
> ```

We strongly disfavor death tests, as documented failure conditions should generally be exceptions, while undocumented failure conditions need not be tested (and also because google's death-test wrapper often breaks valgrind checks).

I suggest that you either document the zero-length condition as an exception rather than termination and change it to `DRAKE_THROW_UNLESS` or else remove both the documentation reference and the death test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpzYEjobMRTc4aDekHU:btehs4m)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
:thumbsup: 

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-:-KpzyrbzI_VkNt-dmZ8f:b4c9lyr)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[drake/automotive/maliput/rndf/test/spline_helpers_test.cc, line 519 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6569#-KpzWoCzPazTlFJO87ed:-Kq-0PVfrTI6SYhBnzVV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/380660a03d5fa529bebfeccdbaf63f9189ac96ff/drake/automotive/maliput/rndf/test/spline_helpers_test.cc#L519)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

We strongly disfavor death tests, as documented failure conditions should generally be exceptions, while undocumented failure conditions need not be tested (and also because google's death-test wrapper often breaks valgrind checks).

I suggest that you either document the zero-length condition as an exception rather than termination and change it to `DRAKE_THROW_UNLESS` or else remove both the documentation reference and the death test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6569)*
<!-- Sent from Reviewable.io -->
